clients/src/main/java/org/apache/kafka/common/memory/SimpleMemoryPool.java

```diff
@@ -0,0 +1,139 @@
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * a simple pool implementation. this implementation just provides a limit on the total outstanding memory.
+ * any buffer allocated must be release()ed always otherwise memory is not marked as reclaimed (and "leak"s)
+ */
+public class SimpleMemoryPool implements MemoryPool {
+    protected final Logger log = LoggerFactory.getLogger(getClass()); //subclass-friendly
+
+    protected final long sizeBytes;
+    protected final boolean strict;
+    protected final long lowMemThreshold;
+    protected final AtomicLong availableMemory;
+    protected final int maxSingleAllocationSize;
+    protected final AtomicLong startOfNonMemPeriod = new AtomicLong();
+    protected Sensor oomTimeSensor;
+
+    public SimpleMemoryPool(long sizeBytes, int maxSingleAllocationSize, boolean strict) {
+        if (sizeBytes <= 0 || maxSingleAllocationSize <= 0 || maxSingleAllocationSize > sizeBytes)
+            throw new IllegalArgumentException("must provide a positive size and max single allocation size smaller than size."
+                + "provided " + sizeBytes + " and " + maxSingleAllocationSize + " respectively");
+        this.sizeBytes = sizeBytes;
+        this.strict = strict;
+        this.lowMemThreshold = (long) (0.1 * this.sizeBytes);
+        this.availableMemory = new AtomicLong(sizeBytes);
+        this.maxSingleAllocationSize = maxSingleAllocationSize;
+    }
+
+    @Override
+    public ByteBuffer tryAllocate(int sizeBytes) {
+        if (sizeBytes < 1)
+            throw new IllegalArgumentException("requested size <=0: " + sizeBytes);
+        if (sizeBytes > maxSingleAllocationSize)
+            throw new IllegalArgumentException("requested size " + sizeBytes + " is larger than maxSingleAllocationSize " + maxSingleAllocationSize);
+
+        long available;
+        boolean success = false;
+        //in strict mode we will only allocate memory if we have at least the size required.
+        //in non-strict mode we will allocate memory if we have _any_ memory available (so available memory
+        //can dip into the negative and max allocated memory would be sizeBytes + maxSingleAllocationSize)
+        long threshold = strict ? sizeBytes : 1;
+        while ((available = availableMemory.get()) >= threshold) {
+            success = availableMemory.compareAndSet(available, available - sizeBytes);
+            if (success)
+                break;
+        }
+
+        if (success) {
+            if (oomTimeSensor != null) {
+                long startOfDrySpell = startOfNonMemPeriod.getAndSet(0);
+                if (startOfDrySpell != 0) {
+                    //how long were we refusing allocation requests for
+                    oomTimeSensor.record(System.currentTimeMillis() - startOfDrySpell);
```

Comment:
i dont understand. the sensor is updated on every single tryAllocate call - successful or not.
only way for the sensor to stop being updated is if the server id idle, in which case there should be plenty of memory available?

if you want better accuracy i could update the sensor when calling release() - this by definition means we have memory, so i could zero-out the OOM time